src/llvm-propagate-addrspaces.cpp

```diff
@@ -74,67 +77,136 @@ static bool isSpecialAS(unsigned AS) {
+        Value *CurrentV = Worklist.back();
+        Worklist.pop_back();
+        for (Value *User : CurrentV->users()) {
+            if (Visited.count(User))
+                continue;
+            Visited.insert(CurrentV);
+            Worklist.push_back(User);
+        }
+    }
+}
+
 Value *PropagateJuliaAddrspaces::LiftPointer(Value *V, Type *LocTy, Instruction *InsertPt) {
     SmallVector<Value *, 4> Stack;
-    Value *CurrentV = V;
+    std::vector<Value *> Worklist;
+    std::set<Value *> LocalVisited;
+    Worklist.push_back(V);
     // Follow pointer casts back, see if we're based on a pointer in
     // an untracked address space, in which case we're allowed to drop
     // intermediate addrspace casts.
-    while (true) {
-        Stack.push_back(CurrentV);
-        if (isa<BitCastInst>(CurrentV))
-            CurrentV = cast<BitCastInst>(CurrentV)->getOperand(0);
-        else if (isa<AddrSpaceCastInst>(CurrentV)) {
-            CurrentV = cast<AddrSpaceCastInst>(CurrentV)->getOperand(0);
-            if (!isSpecialAS(getValueAddrSpace(CurrentV)))
-                break;
+    while (!Worklist.empty()) {
+        Value *CurrentV = Worklist.back();
+        Worklist.pop_back();
+        if (LocalVisited.count(CurrentV)) {
+            continue;
         }
-        else if (isa<GetElementPtrInst>(CurrentV)) {
-            if (LiftingMap.count(CurrentV)) {
-                CurrentV = LiftingMap[CurrentV];
+        while (true) {
+            if (isa<BitCastInst>(CurrentV))
+                CurrentV = cast<BitCastInst>(CurrentV)->getOperand(0);
+            else if (isa<AddrSpaceCastInst>(CurrentV)) {
+                CurrentV = cast<AddrSpaceCastInst>(CurrentV)->getOperand(0);
+                if (!isSpecialAS(getValueAddrSpace(CurrentV)))
+                    break;
+            }
+            else if (isa<GetElementPtrInst>(CurrentV)) {
+                if (LiftingMap.count(CurrentV)) {
+                    CurrentV = LiftingMap[CurrentV];
+                    break;
+                } else if (Visited.count(CurrentV)) {
+                    return nullptr;
+                }
+                Stack.push_back(CurrentV);
+                LocalVisited.insert(CurrentV);
+                CurrentV = cast<GetElementPtrInst>(CurrentV)->getOperand(0);
+            } else if (isa<PHINode>(CurrentV)) {
+                if (LiftingMap.count(CurrentV)) {
+                    break;
+                }
+                for (Value *Incoming : cast<PHINode>(CurrentV)->incoming_values()) {
+                    Worklist.push_back(Incoming);
+                }
+                Stack.push_back(CurrentV);
+                break;
+            } else {
+                // Ok, we've reached a leaf - check if it is eligible for lifting
+                if (!CurrentV->getType()->isPointerTy() ||
+                    isSpecialAS(getValueAddrSpace(CurrentV))) {
+                    // If not, poison all (recursive) users of this value, to prevent
```

Comment:
Seems overly aggressive. It'd be ok to use this value in many contexts (most significantly, as one of the two memory arguments to `memcpy`)

Reply:
Yes, this only poisons for purposes of lifting to addrspace(0). We can't do that if we don't know where it came from, independent of what other transformations are applied to afterwards.

Reply:
Then why does the comment say it poisoned the users to avoid looking at them again?